pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +179,187 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	/// function emits the `TransferApproved` event.
+	///
+	/// This function reserves the necessary deposit from the owner account. If an approval already
+	/// exists, additional funds are reserved only if the updated deposit exceeds the currently
+	/// reserved amount. The existing approval's deadline is also updated.
+	///
+	/// - `owner`: The owner of the collection items.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that will be allowed to take control of the collection items.
+	/// - `maybe_deadline`: The optional deadline (in block numbers) specifying the time limit for
+	///   the approval.
+	pub(crate) fn do_approve_collection_transfer(
+		owner: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+		maybe_deadline: Option<BlockNumberFor<T>>,
+	) -> DispatchResult {
+		ensure!(
+			Self::is_pallet_feature_enabled(PalletFeature::Approvals),
+			Error::<T, I>::MethodDisabled
+		);
+		ensure!(AccountBalance::<T, I>::get(collection, &owner) > 0, Error::<T, I>::NoItemOwned);
+
+		let collection_config = Self::get_collection_config(&collection)?;
+		ensure!(
+			collection_config.is_setting_enabled(CollectionSetting::TransferableItems),
+			Error::<T, I>::ItemsNonTransferable
+		);
+		let now = frame_system::Pallet::<T>::block_number();
+		let deadline = maybe_deadline.map(|d| d.saturating_add(now));
+
+		CollectionApprovals::<T, I>::try_mutate_exists(
+			(&collection, &owner, &delegate),
+			|maybe_approval| -> DispatchResult {
+				let deposit_required = T::CollectionApprovalDeposit::get();
+				let mut current_deposit =
+					maybe_approval.take().map(|(_, deposit)| deposit).unwrap_or_default();
+				if current_deposit < deposit_required {
+					T::Currency::reserve(&owner, deposit_required - current_deposit)?;
+					current_deposit = deposit_required;
+				}
+				*maybe_approval = Some((deadline, current_deposit));
+				Ok(())
+			},
+		)?;
+
+		Self::deposit_event(Event::TransferApproved {
+			collection,
+			item: None,
+			owner,
+			delegate,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	/// Cancels the transfer of all `collection` items of `owner` to a `delegate`.
+	///
+	/// This function is used to cancel the approval for the transfer of the collection items of
+	/// `owner` to a `delegate`. After canceling the approval, the deposit is returned to the
+	/// `owner` account and the `ApprovalCancelled` event is emitted.
+	///
+	/// - `owner`: The owner of the collection items.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that had permission to transfer collection items.
+	pub(crate) fn do_cancel_collection_approval(
+		owner: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+	) -> DispatchResult {
+		let (_, deposit) = CollectionApprovals::<T, I>::take((&collection, &owner, &delegate))
+			.ok_or(Error::<T, I>::Unapproved)?;
+
+		T::Currency::unreserve(&owner, deposit);
```

Comment:
pallet-balances https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/balances/src/impl_currency.rs#L562 shows that it simply returns zero instead of throwing error. But because `Currency` is not only limited to pallet balances, it would be better to only call this if deposit is more than zero.